In [31]:
import pandas as pd
import numpy as np
import matplotlib as plt
import seaborn as sns

In [32]:
%matplotlib inline
pd.set_option('display.max_columns', None)

In [33]:
# import excel file Nov2020PrecinctDetail.xlsx
tn_2020=pd.read_excel('../data/2020_election/Nov2020PrecinctDetail.xlsx',
                      usecols=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE','RNAME1','PARTY1','PVTALLY1',
                             'RNAME2','PARTY2','PVTALLY2','RNAME3','PARTY3','PVTALLY3','RNAME4','PARTY4','PVTALLY4',
                             'RNAME5','PARTY5','PVTALLY5','RNAME6','PARTY6','PVTALLY6','RNAME7','PARTY7','PVTALLY7',
                             'RNAME8','PARTY8','PVTALLY8','RNAME9','PARTY9','PVTALLY9','RNAME10','PARTY10','PVTALLY10'])
tn_2020

COUNTY       PRECINCT  \
0      Anderson  Andersonville   
1      Anderson  Andersonville   
2      Anderson  Andersonville   
3      Anderson  Andersonville   
4      Anderson  Andersonville   
...         ...            ...   
13124    Wilson           25-2   
13125    Wilson           25-2   
13126    Wilson           25-2   
13127    Wilson           25-2   
13128    Wilson           25-2   

                                              OFFICENAME   ELECTDATE  \
0                                United States President  2020-11-03   
1                                United States President  2020-11-03   
2                                   United States Senate  2020-11-03   
3                                   United States Senate  2020-11-03   
4      United States House of Representatives District 3  2020-11-03   
...                                                  ...         ...   
13124                               United States Senate  2020-11-03   
13125                               United States Senate  2020-11-03   
13126  United States House of Representatives District 6  2020-11-03   
13127     Tennessee House of Representatives District 46  2020-11-03   
13128     Tennessee House of Representatives District 57  2020-11-03   

                         RNAME1       PARTY1  PVTALLY1  \
0               Donald J. Trump   Republican      1554   
1      Write-In - Brian Carroll          NaN         0   
2                  Bill Hagerty   Republican      1529   
3       Eric William Stansberry  Independent         4   
4             Chuck Fleischmann   Republican      1507   
...                         ...          ...       ...   
13124              Bill Hagerty   Republican      1224   
13125   Eric William Stansberry  Independent         6   
13126                 John Rose   Republican      1216   
13127                Clark Boyd   Republican       857   
13128              Susan M Lynn   Republican       486   

                          RNAME2       PARTY2  PVTALLY2  \
0                Joseph R. Biden   Democratic     307.0   
1        Write-In - Tom Hoefling          NaN       0.0   
2              Marquita Bradshaw   Democratic     268.0   
3      Write-In - John A. Gentry          NaN       0.0   
4                     Meg Gorman   Democratic     265.0   
...                          ...          ...       ...   
13124          Marquita Bradshaw   Democratic     440.0   
13125  Write-In - John A. Gentry          NaN       0.0   
13126  Christopher Martin Finley   Democratic     427.0   
13127                        NaN          NaN       NaN   
13128                 Tom Sottek  Independent     128.0   

                          RNAME3       PARTY3  PVTALLY3  \
0                Don Blankenship  Independent       1.0   
1        Write-In - Jade Simmons          NaN       0.0   
2      Yomi "Fapas" Faparusi Sr.  Independent       2.0   
3            Write-In - Al Green          NaN       0.0   
4                Amber L. Hysell  Independent      17.0   
...                          ...          ...       ...   
13124  Yomi "Fapas" Faparusi Sr.  Independent       5.0   
13125        Write-In - Al Green          NaN       0.0   
13126      Christopher B. Monday  Independent      47.0   
13127                        NaN          NaN       NaN   
13128                        NaN          NaN       NaN   

                           RNAME4       PARTY4  PVTALLY4  \
0      Roque "Rocky" De La Fuente  Independent       0.0   
1          Write-In - Kasey Wells          NaN       0.0   
2             Jeffrey Alan Grunau  Independent       3.0   
3                             NaN          NaN       NaN   
4          Keith Douglas Sweitzer  Independent      12.0   
...                           ...          ...       ...   
13124         Jeffrey Alan Grunau  Independent       2.0   
13125                         NaN          NaN       NaN   
13126                         NaN          NaN       NaN   
13127               

In [37]:
tn_2020.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13129 entries, 0 to 13128
Data columns (total 34 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   COUNTY      13129 non-null  object 
 1   PRECINCT    13129 non-null  object 
 2   OFFICENAME  13129 non-null  object 
 3   ELECTDATE   13129 non-null  object 
 4   RNAME1      13129 non-null  object 
 5   PARTY1      11164 non-null  object 
 6   PVTALLY1    13129 non-null  int64  
 7   RNAME2      11742 non-null  object 
 8   PARTY2      7504 non-null   object 
 9   PVTALLY2    11742 non-null  float64
 10  RNAME3      9738 non-null   object 
 11  PARTY3      5519 non-null   object 
 12  PVTALLY3    9738 non-null   float64
 13  RNAME4      7387 non-null   object 
 14  PARTY4      4854 non-null   object 
 15  PVTALLY4    7387 non-null   float64
 16  RNAME5      4511 non-null   object 
 17  PARTY5      3927 non-null   object 
 18  PVTALLY5    4511 non-null   float64
 19  RNAME6      3927 non-null

In [38]:
# merge entries in all RNAME columns into new column "name"
tn_2020_namemelt=tn_2020.melt(id_vars=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE'],value_vars=['RNAME1','RNAME2','RNAME3','RNAME4','RNAME5','RNAME6','RNAME7','RNAME8','RNAME9','RNAME10'],value_name="name")
# drop column "variable" created by .melt
tn_2020_namemelt=tn_2020_namemelt.drop(['variable'],axis=1)                          
tn_2020_namemelt.head(9405)

COUNTY             PRECINCT  \
0      Anderson        Andersonville   
1      Anderson        Andersonville   
2      Anderson        Andersonville   
3      Anderson        Andersonville   
4      Anderson        Andersonville   
...         ...                  ...   
9400  Robertson  01-1 East Robertson   
9401  Robertson  01-1 East Robertson   
9402  Robertson         02-1 Woodall   
9403  Robertson         02-1 Woodall   
9404  Robertson         02-1 Woodall   

                                             OFFICENAME   ELECTDATE  \
0                               United States President  2020-11-03   
1                               United States President  2020-11-03   
2                                  United States Senate  2020-11-03   
3                                  United States Senate  2020-11-03   
4     United States House of Representatives District 3  2020-11-03   
...                                                 ...         ...   
9400  United States House of Representatives District 6  2020-11-03   
9401     Tennessee House of Representatives District 66  2020-11-03   
9402                            United States President  2020-11-03   
9403                            United States President  2020-11-03   
9404                               United States Senate  2020-11-03   

                          name  
0              Donald J. Trump  
1     Write-In - Brian Carroll  
2                 Bill Hagerty  
3      Eric William Stansberry  
4            Chuck Fleischmann  
...                        ...  
9400                 John Rose  
9401          Sabi (Doc) Kumar  
9402           Donald J. Trump  
9403  Write-In - Brian Carroll  
9404              Bill Hagerty  

[9405 rows x 5 columns]

In [39]:
# merge entries in all PARTY columns into new column "party"
tn_2020_partymelt=tn_2020.melt(id_vars=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE'],value_vars=['PARTY1','PARTY2','PARTY3','PARTY4','PARTY5','PARTY6','PARTY7','PARTY8','PARTY9','PARTY10'],value_name="party")
# drop column "variable" created by .melt
tn_2020_partymelt=tn_2020_partymelt.drop(['variable'],axis=1)
tn_2020_partymelt.head(9405)

COUNTY             PRECINCT  \
0      Anderson        Andersonville   
1      Anderson        Andersonville   
2      Anderson        Andersonville   
3      Anderson        Andersonville   
4      Anderson        Andersonville   
...         ...                  ...   
9400  Robertson  01-1 East Robertson   
9401  Robertson  01-1 East Robertson   
9402  Robertson         02-1 Woodall   
9403  Robertson         02-1 Woodall   
9404  Robertson         02-1 Woodall   

                                             OFFICENAME   ELECTDATE  \
0                               United States President  2020-11-03   
1                               United States President  2020-11-03   
2                                  United States Senate  2020-11-03   
3                                  United States Senate  2020-11-03   
4     United States House of Representatives District 3  2020-11-03   
...                                                 ...         ...   
9400  United States House of Representatives District 6  2020-11-03   
9401     Tennessee House of Representatives District 66  2020-11-03   
9402                            United States President  2020-11-03   
9403                            United States President  2020-11-03   
9404                               United States Senate  2020-11-03   

            party  
0      Republican  
1             NaN  
2      Republican  
3     Independent  
4      Republican  
...           ...  
9400   Republican  
9401   Republican  
9402   Republican  
9403          NaN  
9404   Republican  

[9405 rows x 5 columns]

In [40]:
# merge entries in all PVTALLY columns into new column "vote_count"
tn_2020_pvtallymelt=tn_2020.melt(id_vars=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE'],value_vars=['PVTALLY1','PVTALLY2','PVTALLY3','PVTALLY4','PVTALLY5','PVTALLY6','PVTALLY7','PVTALLY8','PVTALLY9','PVTALLY10'],value_name="vote_count")
# drop column "variable" created by .melt
tn_2020_pvtallymelt=tn_2020_pvtallymelt.drop(['variable'],axis=1)
tn_2020_pvtallymelt.head(9405)

COUNTY             PRECINCT  \
0      Anderson        Andersonville   
1      Anderson        Andersonville   
2      Anderson        Andersonville   
3      Anderson        Andersonville   
4      Anderson        Andersonville   
...         ...                  ...   
9400  Robertson  01-1 East Robertson   
9401  Robertson  01-1 East Robertson   
9402  Robertson         02-1 Woodall   
9403  Robertson         02-1 Woodall   
9404  Robertson         02-1 Woodall   

                                             OFFICENAME   ELECTDATE  \
0                               United States President  2020-11-03   
1                               United States President  2020-11-03   
2                                  United States Senate  2020-11-03   
3                                  United States Senate  2020-11-03   
4     United States House of Representatives District 3  2020-11-03   
...                                                 ...         ...   
9400  United States House of Representatives District 6  2020-11-03   
9401     Tennessee House of Representatives District 66  2020-11-03   
9402                            United States President  2020-11-03   
9403                            United States President  2020-11-03   
9404                               United States Senate  2020-11-03   

      vote_count  
0         1554.0  
1            0.0  
2         1529.0  
3            4.0  
4         1507.0  
...          ...  
9400      2202.0  
9401      2153.0  
9402      2366.0  
9403         0.0  
9404      2328.0  

[9405 rows x 5 columns]

In [41]:
# add "party" and "vote_count" columns to tn_2020_namemelt
party=tn_2020_partymelt['party']
pvtally=tn_2020_pvtallymelt['vote_count']
tn_2020_namemelt=tn_2020_namemelt.join(party).join(pvtally)
tn_2020_namemelt.head(3)

COUNTY       PRECINCT               OFFICENAME   ELECTDATE  \
0  Anderson  Andersonville  United States President  2020-11-03   
1  Anderson  Andersonville  United States President  2020-11-03   
2  Anderson  Andersonville     United States Senate  2020-11-03   

                       name       party  vote_count  
0           Donald J. Trump  Republican      1554.0  
1  Write-In - Brian Carroll         NaN         0.0  
2              Bill Hagerty  Republican      1529.0

In [42]:
tn_2020_namemelt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 131290 entries, 0 to 131289
Data columns (total 7 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   COUNTY      131290 non-null  object 
 1   PRECINCT    131290 non-null  object 
 2   OFFICENAME  131290 non-null  object 
 3   ELECTDATE   131290 non-null  object 
 4   name        66142 non-null   object 
 5   party       50638 non-null   object 
 6   vote_count  66142 non-null   float64
dtypes: float64(1), object(6)
memory usage: 7.0+ MB


In [43]:
# lower case column names
tn_2020_namemelt.columns=tn_2020_namemelt.columns.str.lower()
tn_2020_namemelt.head(2)

county       precinct               officename   electdate  \
0  Anderson  Andersonville  United States President  2020-11-03   
1  Anderson  Andersonville  United States President  2020-11-03   

                       name       party  vote_count  
0           Donald J. Trump  Republican      1554.0  
1  Write-In - Brian Carroll         NaN         0.0

In [44]:
# change null values in "party" column to "no party"
tn_2020_namemelt['party'].fillna("No party", inplace = True)
tn_2020_namemelt.head(2)

county       precinct               officename   electdate  \
0  Anderson  Andersonville  United States President  2020-11-03   
1  Anderson  Andersonville  United States President  2020-11-03   

                       name       party  vote_count  
0           Donald J. Trump  Republican      1554.0  
1  Write-In - Brian Carroll    No party         0.0

In [45]:
tn_2020_namemelt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 131290 entries, 0 to 131289
Data columns (total 7 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   county      131290 non-null  object 
 1   precinct    131290 non-null  object 
 2   officename  131290 non-null  object 
 3   electdate   131290 non-null  object 
 4   name        66142 non-null   object 
 5   party       131290 non-null  object 
 6   vote_count  66142 non-null   float64
dtypes: float64(1), object(6)
memory usage: 7.0+ MB


In [46]:
# change null values in "vote_count" column to "0"
tn_2020_namemelt['vote_count'].fillna(0)

0         1554.0
1            0.0
2         1529.0
3            4.0
4         1507.0
           ...  
131285      10.0
131286       0.0
131287       0.0
131288       0.0
131289       0.0
Name: vote_count, Length: 131290, dtype: float64

In [47]:
tn_2020_namemelt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 131290 entries, 0 to 131289
Data columns (total 7 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   county      131290 non-null  object 
 1   precinct    131290 non-null  object 
 2   officename  131290 non-null  object 
 3   electdate   131290 non-null  object 
 4   name        66142 non-null   object 
 5   party       131290 non-null  object 
 6   vote_count  66142 non-null   float64
dtypes: float64(1), object(6)
memory usage: 7.0+ MB


# 66142 non-null names and 66142 non-null vote counts
# Do the nulls in "name" correlate one to one with "no party"?

In [48]:
tn_2020_namemelt[tn_2020_namemelt['name'].isna()]

county      precinct  \
13140   Anderson    Briceville   
13146   Anderson      Bull Run   
13152   Anderson       Clinton   
13158   Anderson  Clinton High   
13164   Anderson       Claxton   
...          ...           ...   
131284    Wilson          25-2   
131286    Wilson          25-2   
131287    Wilson          25-2   
131288    Wilson          25-2   
131289    Wilson          25-2   

                                               officename   electdate name  \
13140      Tennessee House of Representatives District 33  2020-11-03  NaN   
13146      Tennessee House of Representatives District 33  2020-11-03  NaN   
13152      Tennessee House of Representatives District 33  2020-11-03  NaN   
13158      Tennessee House of Representatives District 33  2020-11-03  NaN   
13164      Tennessee House of Representatives District 33  2020-11-03  NaN   
...                                                   ...         ...  ...   
131284                            United States President  2020-11-03  NaN   
131286                               United States Senate  2020-11-03  NaN   
131287  United States House of Representatives District 6  2020-11-03  NaN   
131288     Tennessee House of Representatives District 46  2020-11-03  NaN   
131289     Tennessee House of Representatives District 57  2020-11-03  NaN   

           party  vote_count  
13140   No party         NaN  
13146   No party         NaN  
13152   No party         NaN  
13158   No party         NaN  
13164   No party         NaN  
...          ...         ...  
131284  No party         NaN  
131286  No party         NaN  
131287  No party         NaN  
131288  No party         NaN  
131289  No party         NaN  

[65148 rows x 7 columns]

In [49]:
tn_2020_namemelt[tn_2020_namemelt['name'].isna()].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 65148 entries, 13140 to 131289
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   county      65148 non-null  object 
 1   precinct    65148 non-null  object 
 2   officename  65148 non-null  object 
 3   electdate   65148 non-null  object 
 4   name        0 non-null      object 
 5   party       65148 non-null  object 
 6   vote_count  0 non-null      float64
dtypes: float64(1), object(6)
memory usage: 4.0+ MB


### Looks safe to drop rows with no votes tallied

In [50]:
# drop rows with no name and vote count
tn_2020_final=tn_2020_namemelt.dropna()

In [51]:
tn_2020_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 66142 entries, 0 to 131285
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   county      66142 non-null  object 
 1   precinct    66142 non-null  object 
 2   officename  66142 non-null  object 
 3   electdate   66142 non-null  object 
 4   name        66142 non-null  object 
 5   party       66142 non-null  object 
 6   vote_count  66142 non-null  float64
dtypes: float64(1), object(6)
memory usage: 4.0+ MB


In [52]:
tn_2020_final

county       precinct  \
0       Anderson  Andersonville   
1       Anderson  Andersonville   
2       Anderson  Andersonville   
3       Anderson  Andersonville   
4       Anderson  Andersonville   
...          ...            ...   
131273    Wilson           24-1   
131277    Wilson           25-1   
131279    Wilson           25-1   
131283    Wilson           25-2   
131285    Wilson           25-2   

                                               officename   electdate  \
0                                 United States President  2020-11-03   
1                                 United States President  2020-11-03   
2                                    United States Senate  2020-11-03   
3                                    United States Senate  2020-11-03   
4       United States House of Representatives District 3  2020-11-03   
...                                                   ...         ...   
131273                               United States Senate  2020-11-03   
131277                            United States President  2020-11-03   
131279                               United States Senate  2020-11-03   
131283                            United States President  2020-11-03   
131285                               United States Senate  2020-11-03   

                            name        party  vote_count  
0                Donald J. Trump   Republican      1554.0  
1       Write-In - Brian Carroll     No party         0.0  
2                   Bill Hagerty   Republican      1529.0  
3        Eric William Stansberry  Independent         4.0  
4              Chuck Fleischmann   Republican      1507.0  
...                          ...          ...         ...  
131273              Kacey Morgan  Independent        14.0  
131277     Write-In - R19 Boddie     No party         0.0  
131279              Kacey Morgan  Independent        14.0  
131283     Write-In - R19 Boddie     No party         0.0  
131285              Kacey Morgan  Independent        10.0  

[66142 rows x 7 columns]

In [53]:
# convert vote_count data from float64 to int 32 to to drop ".0"
tn_2020_final=tn_2020_final.astype({'vote_count':'int32'})
tn_2020_final

county       precinct  \
0       Anderson  Andersonville   
1       Anderson  Andersonville   
2       Anderson  Andersonville   
3       Anderson  Andersonville   
4       Anderson  Andersonville   
...          ...            ...   
131273    Wilson           24-1   
131277    Wilson           25-1   
131279    Wilson           25-1   
131283    Wilson           25-2   
131285    Wilson           25-2   

                                               officename   electdate  \
0                                 United States President  2020-11-03   
1                                 United States President  2020-11-03   
2                                    United States Senate  2020-11-03   
3                                    United States Senate  2020-11-03   
4       United States House of Representatives District 3  2020-11-03   
...                                                   ...         ...   
131273                               United States Senate  2020-11-03   
131277                            United States President  2020-11-03   
131279                               United States Senate  2020-11-03   
131283                            United States President  2020-11-03   
131285                               United States Senate  2020-11-03   

                            name        party  vote_count  
0                Donald J. Trump   Republican        1554  
1       Write-In - Brian Carroll     No party           0  
2                   Bill Hagerty   Republican        1529  
3        Eric William Stansberry  Independent           4  
4              Chuck Fleischmann   Republican        1507  
...                          ...          ...         ...  
131273              Kacey Morgan  Independent          14  
131277     Write-In - R19 Boddie     No party           0  
131279              Kacey Morgan  Independent          14  
131283     Write-In - R19 Boddie     No party           0  
131285              Kacey Morgan  Independent          10  

[66142 rows x 7 columns]

In [55]:
tn_2020_final.dtypes

county        object
precinct      object
officename    object
electdate     object
name          object
party         object
vote_count     int32
dtype: object

In [56]:
tn_2020_final['officename'].value_counts()

United States President                              27510
United States Senate                                 25506
United States House of Representatives District 3     1395
United States House of Representatives District 7     1100
United States House of Representatives District 8      976
                                                     ...  
Tennessee House of Representatives District 45          14
Tennessee House of Representatives District 99          12
Tennessee House of Representatives District 1           11
Tennessee House of Representatives District 68          11
Tennessee House of Representatives District 81           9
Name: officename, Length: 126, dtype: int64

In [58]:
tn_2020_final['officename'].unique()

array(['United States President', 'United States Senate',
       'United States House of Representatives District 3',
       'Tennessee House of Representatives District 36',
       'Tennessee House of Representatives District 33',
       'United States House of Representatives District 4',
       'Tennessee Senate District 14',
       'Tennessee House of Representatives District 62',
       'United States House of Representatives District 7',
       'Tennessee Senate District 24',
       'Tennessee House of Representatives District 75',
       'United States House of Representatives District 8',
       'Tennessee House of Representatives District 31',
       'United States House of Representatives District 2',
       'Tennessee Senate District 2',
       'Tennessee House of Representatives District 8',
       'Tennessee House of Representatives District 20',
       'Tennessee Senate District 10',
       'Tennessee House of Representatives District 24',
       'Tennessee House of Repre

In [59]:
tn_2020_final['vote_count'].value_counts()

0       22571
1        5980
2        4279
3        3195
4        2337
        ...  
2586        1
3382        1
4889        1
4430        1
2918        1
Name: vote_count, Length: 2726, dtype: int64

## 22571 rows have a vote count of 0. Drop those rows

In [60]:
tn_2020_final=tn_2020_final[tn_2020_final['vote_count'] != 0]
tn_2020_final=tn_2020_final.reset_index(drop=True)
tn_2020_final.head()

county       precinct                                         officename  \
0  Anderson  Andersonville                            United States President   
1  Anderson  Andersonville                               United States Senate   
2  Anderson  Andersonville                               United States Senate   
3  Anderson  Andersonville  United States House of Representatives District 3   
4  Anderson  Andersonville     Tennessee House of Representatives District 36   

    electdate                     name        party  vote_count  
0  2020-11-03          Donald J. Trump   Republican        1554  
1  2020-11-03             Bill Hagerty   Republican        1529  
2  2020-11-03  Eric William Stansberry  Independent           4  
3  2020-11-03        Chuck Fleischmann   Republican        1507  
4  2020-11-03            Dennis Powers   Republican        1528

In [61]:
# confirm no 0's in vote count
tn_2020_final['vote_count'].value_counts()

1       5980
2       4279
3       3195
4       2337
5       1869
        ... 
2322       1
2386       1
2983       1
2418       1
2049       1
Name: vote_count, Length: 2725, dtype: int64

In [62]:
# create data frame tn_2020_pres for presidental election
tn_2020_pres_int=tn_2020_final[tn_2020_final['officename'].str.match('United States President')]
tn_2020_pres_int=tn_2020_pres_int.reset_index(drop=True)
tn_2020_pres_int

county       precinct               officename   electdate  \
0      Anderson  Andersonville  United States President  2020-11-03   
1      Anderson     Briceville  United States President  2020-11-03   
2      Anderson       Bull Run  United States President  2020-11-03   
3      Anderson        Clinton  United States President  2020-11-03   
4      Anderson   Clinton High  United States President  2020-11-03   
...         ...            ...                      ...         ...   
13557    Wilson           23-1  United States President  2020-11-03   
13558    Wilson           24-1  United States President  2020-11-03   
13559    Wilson           25-1  United States President  2020-11-03   
13560    Wilson           25-2  United States President  2020-11-03   
13561  Davidson           06-3  United States President  2020-11-03   

                        name        party  vote_count  
0            Donald J. Trump   Republican        1554  
1            Donald J. Trump   Republican         333  
2            Donald J. Trump   Republican        1692  
3            Donald J. Trump   Republican        1548  
4            Donald J. Trump   Republican         962  
...                      ...          ...         ...  
13557             Kanye West  Independent          14  
13558             Kanye West  Independent          19  
13559             Kanye West  Independent          24  
13560             Kanye West  Independent          10  
13561  Write-In - R19 Boddie     No party           1  

[13562 rows x 7 columns]

In [63]:
tn_2020_pres_int.head(20)

county        precinct               officename   electdate  \
0   Anderson   Andersonville  United States President  2020-11-03   
1   Anderson      Briceville  United States President  2020-11-03   
2   Anderson        Bull Run  United States President  2020-11-03   
3   Anderson         Clinton  United States President  2020-11-03   
4   Anderson    Clinton High  United States President  2020-11-03   
5   Anderson         Claxton  United States President  2020-11-03   
6   Anderson  Clinton Middle  United States President  2020-11-03   
7   Anderson    Dutch Valley  United States President  2020-11-03   
8   Anderson    Emory Valley  United States President  2020-11-03   
9   Anderson    Emory Valley  United States President  2020-11-03   
10  Anderson        Fairview  United States President  2020-11-03   
11  Anderson     Glen Alpine  United States President  2020-11-03   
12  Anderson        Glenwood  United States President  2020-11-03   
13  Anderson        Glenwood  United States President  2020-11-03   
14  Anderson   Hendrix Creek  United States President  2020-11-03   
15  Anderson   Hendrix Creek  United States President  2020-11-03   
16  Anderson   Highland View  United States President  2020-11-03   
17  Anderson    Lake City MS  United States President  2020-11-03   
18  Anderson          Marlow  United States President  2020-11-03   
19  Anderson   North Clinton  United States President  2020-11-03   

                        name       party  vote_count  
0            Donald J. Trump  Republican        1554  
1            Donald J. Trump  Republican         333  
2            Donald J. Trump  Republican        1692  
3            Donald J. Trump  Republican        1548  
4            Donald J. Trump  Republican         962  
5            Donald J. Trump  Republican        1637  
6            Donald J. Trump  Republican         533  
7            Donald J. Trump  Republican         344  
8            Donald J. Trump  Republican        1077  
9   Write-In - Brian Carroll    No party           2  
10           Donald J. Trump  Republican         680  
11           Donald J. Trump  Republican        1444  
12           Donald J. Trump  Republican         852  
13  Write-In - Brian Carroll    No party           2  
14           Donald J. Trump  Republican        1006  
15  Write-In - Brian Carroll    No party           1  
16           Donald J. Trump  Republican         403  
17           Donald J. Trump  Republican        1517  
18           Donald J. Trump  Republican         588  
19           Donald J. Trump  Republican        1016

### Add column that has total vote count by count for use in calculating county vote percentages

In [64]:
# group county precincts into county per candidate, in descending order of votes within county
tn_2020_pres_int=tn_2020_pres_int.groupby(['electdate','officename','county','name','party'], as_index=False)['vote_count'].sum()
tn_2020_pres=tn_2020_pres_int.sort_values(by=['county','vote_count'],ascending=False).reset_index(drop=True)
tn_2020_pres.head(20)

electdate               officename      county  \
0   2020-11-03  United States President      Wilson   
1   2020-11-03  United States President      Wilson   
2   2020-11-03  United States President      Wilson   
3   2020-11-03  United States President      Wilson   
4   2020-11-03  United States President      Wilson   
5   2020-11-03  United States President      Wilson   
6   2020-11-03  United States President      Wilson   
7   2020-11-03  United States President      Wilson   
8   2020-11-03  United States President      Wilson   
9   2020-11-03  United States President      Wilson   
10  2020-11-03  United States President      Wilson   
11  2020-11-03  United States President      Wilson   
12  2020-11-03  United States President  Williamson   
13  2020-11-03  United States President  Williamson   
14  2020-11-03  United States President  Williamson   
15  2020-11-03  United States President  Williamson   
16  2020-11-03  United States President  Williamson   
17  2020-11-03  United States President  Williamson   
18  2020-11-03  United States President  Williamson   
19  2020-11-03  United States President  Williamson   

                          name        party  vote_count  
0              Donald J. Trump   Republican       50296  
1              Joseph R. Biden   Democratic       22254  
2                 Jo Jorgensen  Independent         781  
3                   Kanye West  Independent         449  
4              Don Blankenship  Independent         163  
5               Alyson Kennedy  Independent         126  
6               Gloria La Riva  Independent         104  
7                Howie Hawkins  Independent          81  
8   Roque "Rocky" De La Fuente  Independent          51  
9     Write-In - Brian Carroll     No party          19  
10     Write-In - Jade Simmons     No party           3  
11     Write-In - Tom Hoefling     No party           3  
12             Donald J. Trump   Republican       86469  
13             Joseph R. Biden   Democratic       50161  
14                Jo Jorgensen  Independent        1494  
15             Don Blankenship  Independent         276  
16                  Kanye West  Independent         257  
17               Howie Hawkins  Independent         116  
18  Roque "Rocky" De La Fuente  Independent          88  
19    Write-In - Brian Carroll     No party          72

In [65]:
# create new dataframe "county_vote" to use to find sum of votes per county
county_vote=tn_2020_pres.loc[0:,['county','vote_count']]
county_vote

county  vote_count
0      Wilson       50296
1      Wilson       22254
2      Wilson         781
3      Wilson         449
4      Wilson         163
..        ...         ...
893  Anderson          13
894  Anderson          12
895  Anderson          11
896  Anderson           1
897  Anderson           1

[898 rows x 2 columns]

In [66]:
#sum county votes
county_vote_sum=county_vote.groupby(['county'])['vote_count'].sum()
county_vote_sum

county
Anderson       35570
Bedford        19088
Benton          7260
Bledsoe         5758
Blount         66609
               ...  
Wayne           6669
Weakley        13735
White          11895
Williamson    139016
Wilson         74330
Name: vote_count, Length: 95, dtype: int32

In [67]:
# use "county_vote_sum" to add column "sum_county_vote" to tn_2020_pres correlating with value in each row
tn_2020_pres['sum_county_vote']=tn_2020_pres['county'].map(county_vote_sum)
tn_2020_pres.head(20)

electdate               officename      county  \
0   2020-11-03  United States President      Wilson   
1   2020-11-03  United States President      Wilson   
2   2020-11-03  United States President      Wilson   
3   2020-11-03  United States President      Wilson   
4   2020-11-03  United States President      Wilson   
5   2020-11-03  United States President      Wilson   
6   2020-11-03  United States President      Wilson   
7   2020-11-03  United States President      Wilson   
8   2020-11-03  United States President      Wilson   
9   2020-11-03  United States President      Wilson   
10  2020-11-03  United States President      Wilson   
11  2020-11-03  United States President      Wilson   
12  2020-11-03  United States President  Williamson   
13  2020-11-03  United States President  Williamson   
14  2020-11-03  United States President  Williamson   
15  2020-11-03  United States President  Williamson   
16  2020-11-03  United States President  Williamson   
17  2020-11-03  United States President  Williamson   
18  2020-11-03  United States President  Williamson   
19  2020-11-03  United States President  Williamson   

                          name        party  vote_count  sum_county_vote  
0              Donald J. Trump   Republican       50296            74330  
1              Joseph R. Biden   Democratic       22254            74330  
2                 Jo Jorgensen  Independent         781            74330  
3                   Kanye West  Independent         449            74330  
4              Don Blankenship  Independent         163            74330  
5               Alyson Kennedy  Independent         126            74330  
6               Gloria La Riva  Independent         104            74330  
7                Howie Hawkins  Independent          81            74330  
8   Roque "Rocky" De La Fuente  Independent          51            74330  
9     Write-In - Brian Carroll     No party          19            74330  
10     Write-In - Jade Simmons     No party           3            74330  
11     Write-In - Tom Hoefling     No party           3            74330  
12             Donald J. Trump   Republican       86469           139016  
13             Joseph R. Biden   Democratic       50161           139016  
14                Jo Jorgensen  Independent        1494           139016  
15             Don Blankenship  Independent         276           139016  
16                  Kanye West  Independent         257           139016  
17               Howie Hawkins  Independent         116           139016  
18  Roque "Rocky" De La Fuente  Independent          88           139016  
19    Write-In - Brian Carroll     No party          72           139016

In [69]:
# create "pct_county_vote" column for tn_2020_pres
tn_2020_pres['pct_county_vote']=round(tn_2020_pres['vote_count']/tn_2020_pres['sum_county_vote']*100,2)
tn_2020_pres

electdate               officename    county  \
0    2020-11-03  United States President    Wilson   
1    2020-11-03  United States President    Wilson   
2    2020-11-03  United States President    Wilson   
3    2020-11-03  United States President    Wilson   
4    2020-11-03  United States President    Wilson   
..          ...                      ...       ...   
893  2020-11-03  United States President  Anderson   
894  2020-11-03  United States President  Anderson   
895  2020-11-03  United States President  Anderson   
896  2020-11-03  United States President  Anderson   
897  2020-11-03  United States President  Anderson   

                           name        party  vote_count  sum_county_vote  \
0               Donald J. Trump   Republican       50296            74330   
1               Joseph R. Biden   Democratic       22254            74330   
2                  Jo Jorgensen  Independent         781            74330   
3                    Kanye West  Independent         449            74330   
4               Don Blankenship  Independent         163            74330   
..                          ...          ...         ...              ...   
893              Gloria La Riva  Independent          13            35570   
894  Roque "Rocky" De La Fuente  Independent          12            35570   
895    Write-In - Brian Carroll     No party          11            35570   
896     Write-In - Jade Simmons     No party           1            35570   
897     Write-In - Tom Hoefling     No party           1            35570   

     pct_county_vote  
0              67.67  
1              29.94  
2               1.05  
3               0.60  
4               0.22  
..               ...  
893             0.04  
894             0.03  
895             0.03  
896             0.00  
897             0.00  

[898 rows x 8 columns]

In [70]:
# how many parties?
tn_2020_pres['party'].value_counts()

Independent    637
Republican      95
Democratic      95
No party        71
Name: party, dtype: int64

### Create dataframe "tn_2020_county_pres_pct"  with percent values per party per county

In [71]:
# create new dataframe with reduced columns values only
tn_2020_pct_county=tn_2020_pres.loc[0:,['electdate','county','party','pct_county_vote']]
tn_2020_pct_county

electdate    county        party  pct_county_vote
0    2020-11-03    Wilson   Republican            67.67
1    2020-11-03    Wilson   Democratic            29.94
2    2020-11-03    Wilson  Independent             1.05
3    2020-11-03    Wilson  Independent             0.60
4    2020-11-03    Wilson  Independent             0.22
..          ...       ...          ...              ...
893  2020-11-03  Anderson  Independent             0.04
894  2020-11-03  Anderson  Independent             0.03
895  2020-11-03  Anderson     No party             0.03
896  2020-11-03  Anderson     No party             0.00
897  2020-11-03  Anderson     No party             0.00

[898 rows x 4 columns]

In [72]:
# create new dataframe with Republican values only
tn_2020_rep_pct=tn_2020_pct_county[tn_2020_pct_county['party']== "Republican"].reset_index(drop=True)
tn_2020_rep_pct

electdate      county       party  pct_county_vote
0   2020-11-03      Wilson  Republican            67.67
1   2020-11-03  Williamson  Republican            62.20
2   2020-11-03       White  Republican            80.76
3   2020-11-03     Weakley  Republican            75.69
4   2020-11-03       Wayne  Republican            86.89
..         ...         ...         ...              ...
90  2020-11-03      Blount  Republican            71.12
91  2020-11-03     Bledsoe  Republican            82.06
92  2020-11-03      Benton  Republican            78.07
93  2020-11-03     Bedford  Republican            75.20
94  2020-11-03    Anderson  Republican            65.18

[95 rows x 4 columns]

In [75]:
# rename 'pct_county_vote' to 'rep_pct' and drop "party" column to prepare for merge 
tn_2020_rep_pct=tn_2020_rep_pct.rename(columns={'pct_county_vote':'Republican'},inplace=False).drop(['party'], axis=1)
tn_2020_rep_pct

electdate      county  Republican
0   2020-11-03      Wilson       67.67
1   2020-11-03  Williamson       62.20
2   2020-11-03       White       80.76
3   2020-11-03     Weakley       75.69
4   2020-11-03       Wayne       86.89
..         ...         ...         ...
90  2020-11-03      Blount       71.12
91  2020-11-03     Bledsoe       82.06
92  2020-11-03      Benton       78.07
93  2020-11-03     Bedford       75.20
94  2020-11-03    Anderson       65.18

[95 rows x 3 columns]

In [76]:
# create new dataframe with Democratic values only
tn_2020_dem_pct=tn_2020_pct_county[tn_2020_pct_county['party']== "Democratic"].reset_index(drop=True)
tn_2020_dem_pct

electdate      county       party  pct_county_vote
0   2020-11-03      Wilson  Democratic            29.94
1   2020-11-03  Williamson  Democratic            36.08
2   2020-11-03       White  Democratic            18.02
3   2020-11-03     Weakley  Democratic            21.99
4   2020-11-03       Wayne  Democratic            12.30
..         ...         ...         ...              ...
90  2020-11-03      Blount  Democratic            26.92
91  2020-11-03     Bledsoe  Democratic            16.86
92  2020-11-03      Benton  Democratic            21.06
93  2020-11-03     Bedford  Democratic            23.33
94  2020-11-03    Anderson  Democratic            33.01

[95 rows x 4 columns]

In [77]:
# rename 'pct_county_vote' to 'dem_pct' and drop "party" to prepare for merge 
tn_2020_dem_pct=tn_2020_dem_pct.rename(columns={'pct_county_vote':'Democratic'},inplace=False).drop(['party'], axis=1)
tn_2020_dem_pct

electdate      county  Democratic
0   2020-11-03      Wilson       29.94
1   2020-11-03  Williamson       36.08
2   2020-11-03       White       18.02
3   2020-11-03     Weakley       21.99
4   2020-11-03       Wayne       12.30
..         ...         ...         ...
90  2020-11-03      Blount       26.92
91  2020-11-03     Bledsoe       16.86
92  2020-11-03      Benton       21.06
93  2020-11-03     Bedford       23.33
94  2020-11-03    Anderson       33.01

[95 rows x 3 columns]

In [79]:
# create new dataframe with non Republican or Democratic values only
tn_2020_other_pct=tn_2020_pct_county[(tn_2020_pct_county['party']=="Independent")|(tn_2020_pct_county['party']=="No party")].reset_index(drop=True)
tn_2020_other_pct

electdate    county        party  pct_county_vote
0    2020-11-03    Wilson  Independent             1.05
1    2020-11-03    Wilson  Independent             0.60
2    2020-11-03    Wilson  Independent             0.22
3    2020-11-03    Wilson  Independent             0.17
4    2020-11-03    Wilson  Independent             0.14
..          ...       ...          ...              ...
703  2020-11-03  Anderson  Independent             0.04
704  2020-11-03  Anderson  Independent             0.03
705  2020-11-03  Anderson     No party             0.03
706  2020-11-03  Anderson     No party             0.00
707  2020-11-03  Anderson     No party             0.00

[708 rows x 4 columns]

In [80]:
# sum by county
tn_2020_other_pct=tn_2020_other_pct.groupby(['county'],as_index=False)['pct_county_vote'].sum()
tn_2020_other_pct

county  pct_county_vote
0     Anderson             1.81
1      Bedford             1.48
2       Benton             0.86
3      Bledsoe             1.07
4       Blount             1.97
..         ...              ...
90       Wayne             0.79
91     Weakley             2.33
92       White             1.24
93  Williamson             1.70
94      Wilson             2.39

[95 rows x 2 columns]

In [81]:
# rename 'pct_county_vote' to 'other_pct' and drop "party" to prepare for merge 
tn_2020_other_pct=tn_2020_other_pct.rename(columns={'pct_county_vote':'Other'})
tn_2020_other_pct

county  Other
0     Anderson   1.81
1      Bedford   1.48
2       Benton   0.86
3      Bledsoe   1.07
4       Blount   1.97
..         ...    ...
90       Wayne   0.79
91     Weakley   2.33
92       White   1.24
93  Williamson   1.70
94      Wilson   2.39

[95 rows x 2 columns]

In [82]:
# join tn_2020_rep_pct, tn_2020_dem_pct and tn_2020_other_pct to form tn_2020_county_pres_pct
dem=tn_2020_dem_pct['Democratic']
other=tn_2020_other_pct['Other']
tn_2020_county_pres_pct=tn_2020_rep_pct.join(dem).join(other)
tn_2020_county_pres_pct

electdate      county  Republican  Democratic  Other
0   2020-11-03      Wilson       67.67       29.94   1.81
1   2020-11-03  Williamson       62.20       36.08   1.48
2   2020-11-03       White       80.76       18.02   0.86
3   2020-11-03     Weakley       75.69       21.99   1.07
4   2020-11-03       Wayne       86.89       12.30   1.97
..         ...         ...         ...         ...    ...
90  2020-11-03      Blount       71.12       26.92   0.79
91  2020-11-03     Bledsoe       82.06       16.86   2.33
92  2020-11-03      Benton       78.07       21.06   1.24
93  2020-11-03     Bedford       75.20       23.33   1.70
94  2020-11-03    Anderson       65.18       33.01   2.39

[95 rows x 5 columns]

### Add column for Republican margin over Democrat for heat mapping in Tableau

In [85]:
tn_2020_county_pres_pct['Rep_Margin']= tn_2020_county_pres_pct['Republican']-tn_2020_county_pres_pct['Democratic']
tn_2020_county_pres_pct.head(40)

electdate      county  Republican  Democratic  Other  Rep_Margin
0   2020-11-03      Wilson       67.67       29.94   1.81       37.73
1   2020-11-03  Williamson       62.20       36.08   1.48       26.12
2   2020-11-03       White       80.76       18.02   0.86       62.74
3   2020-11-03     Weakley       75.69       21.99   1.07       53.70
4   2020-11-03       Wayne       86.89       12.30   1.97       74.59
5   2020-11-03  Washington       67.18       30.96   1.77       36.22
6   2020-11-03      Warren       74.02       24.51   1.07       49.51
7   2020-11-03   Van Buren       80.18       18.62   1.61       61.56
8   2020-11-03       Union       83.75       15.38   1.20       68.37
9   2020-11-03      Unicoi       79.44       19.44   1.53       60.00
10  2020-11-03   Trousdale       73.44       25.31   1.50       48.13
11  2020-11-03      Tipton       73.49       25.04   2.91       48.45
12  2020-11-03      Sumner       68.50       29.88   1.07       38.62
13  2020-11-03    Sullivan       75.12       23.23   1.08       51.89
14  2020-11-03     Stewart       78.62       19.57   1.10       59.05
15  2020-11-03       Smith       78.84       19.91   2.85       58.93
16  2020-11-03      Shelby       33.98       64.42   0.98      -30.44
17  2020-11-03      Sevier       77.60       20.03   1.19       57.57
18  2020-11-03  Sequatchie       80.74       17.90   3.14       62.84
19  2020-11-03       Scott       88.42       10.89   2.07       77.53
20  2020-11-03  Rutherford       56.63       41.24   1.07       15.39
21  2020-11-03   Robertson       72.77       25.78   2.36       46.99
22  2020-11-03       Roane       74.20       23.32   1.04       50.88
23  2020-11-03        Rhea       81.03       17.37   1.16       63.66
24  2020-11-03      Putnam       70.73       27.34   0.84       43.39
25  2020-11-03        Polk       81.24       17.85   1.48       63.39
26  2020-11-03     Pickett       81.24       17.91   1.37       63.33
27  2020-11-03       Perry       80.95       17.94   0.93       63.01
28  2020-11-03     Overton       78.89       20.26   1.01       58.63
29  2020-11-03       Obion       79.80       19.15   2.24       60.65
30  2020-11-03      Morgan       84.22       14.18   1.11       70.04
31  2020-11-03       Moore       81.60       16.19   1.29       65.41
32  2020-11-03  Montgomery       54.96       42.30   2.04       12.66
33  2020-11-03      Monroe       80.70       18.10   0.37       62.60
34  2020-11-03       Meigs       80.75       18.22   1.23       62.53
35  2020-11-03     McNairy       80.65       17.23   1.76       63.42
36  2020-11-03      McMinn       79.66       19.09   1.35       60.57
37  2020-11-03       Maury       67.44       30.90   1.11       36.54
38  2020-11-03    Marshall       74.22       24.23   1.09       49.99
39  2020-11-03      Marion       74.77       23.97   1.73       50.80

In [86]:
#read TN_county_index_shape.csv to add index from TN Counties shape file for joining in Tableau
county_index=pd.read_csv('../data/TN_county_index_shape.csv')
county_index

county  INDEX_NO
0     Anderson        79
1      Bedford        26
2       Benton        41
3      Bledsoe        31
4       Blount        52
..         ...       ...
90       Wayne         6
91     Weakley        56
92       White        67
93  Williamson        44
94      Wilson        65

[95 rows x 2 columns]

In [88]:
# join with tn_2020_county_pres_pct
tn_2020_county_pres_pct=tn_2020_county_pres_pct.merge(county_index,on="county")
tn_2020_county_pres_pct

electdate      county  Republican  Democratic  Other  Rep_Margin  \
0   2020-11-03      Wilson       67.67       29.94   1.81       37.73   
1   2020-11-03  Williamson       62.20       36.08   1.48       26.12   
2   2020-11-03       White       80.76       18.02   0.86       62.74   
3   2020-11-03     Weakley       75.69       21.99   1.07       53.70   
4   2020-11-03       Wayne       86.89       12.30   1.97       74.59   
..         ...         ...         ...         ...    ...         ...   
89  2020-11-03      Blount       71.12       26.92   0.79       44.20   
90  2020-11-03     Bledsoe       82.06       16.86   2.33       65.20   
91  2020-11-03      Benton       78.07       21.06   1.24       57.01   
92  2020-11-03     Bedford       75.20       23.33   1.70       51.87   
93  2020-11-03    Anderson       65.18       33.01   2.39       32.17   

    INDEX_NO  
0         65  
1         44  
2         67  
3         56  
4          6  
..       ...  
89        52  
90        31  
91        41  
92        26  
93        79  

[94 rows x 7 columns]

In [89]:
tn_2020_county_pres_pct.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 94 entries, 0 to 93
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   electdate   94 non-null     object 
 1   county      94 non-null     object 
 2   Republican  94 non-null     float64
 3   Democratic  94 non-null     float64
 4   Other       94 non-null     float64
 5   Rep_Margin  94 non-null     float64
 6   INDEX_NO    94 non-null     int64  
dtypes: float64(4), int64(1), object(2)
memory usage: 5.9+ KB


In [90]:
# covert "electdate" to datetime from object
tn_2020_county_pres_pct['electdate']=pd.to_datetime(tn_2020_county_pres_pct['electdate'],format='%Y-%m-%d')
tn_2020_county_pres_pct

electdate      county  Republican  Democratic  Other  Rep_Margin  INDEX_NO
0  2020-11-03      Wilson       67.67       29.94   1.81       37.73        65
1  2020-11-03  Williamson       62.20       36.08   1.48       26.12        44
2  2020-11-03       White       80.76       18.02   0.86       62.74        67
3  2020-11-03     Weakley       75.69       21.99   1.07       53.70        56
4  2020-11-03       Wayne       86.89       12.30   1.97       74.59         6
..        ...         ...         ...         ...    ...         ...       ...
89 2020-11-03      Blount       71.12       26.92   0.79       44.20        52
90 2020-11-03     Bledsoe       82.06       16.86   2.33       65.20        31
91 2020-11-03      Benton       78.07       21.06   1.24       57.01        41
92 2020-11-03     Bedford       75.20       23.33   1.70       51.87        26
93 2020-11-03    Anderson       65.18       33.01   2.39       32.17        79

[94 rows x 7 columns]

In [91]:
tn_2020_county_pres_pct.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 94 entries, 0 to 93
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   electdate   94 non-null     datetime64[ns]
 1   county      94 non-null     object        
 2   Republican  94 non-null     float64       
 3   Democratic  94 non-null     float64       
 4   Other       94 non-null     float64       
 5   Rep_Margin  94 non-null     float64       
 6   INDEX_NO    94 non-null     int64         
dtypes: datetime64[ns](1), float64(4), int64(1), object(1)
memory usage: 5.9+ KB


In [92]:
tn_2020_county_pres_pct.to_excel(r'../data/2020_election/tn_2020_county_pres_pct.xlsx',index=False)